text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,291 @@
+- Feature Name: generic_dependent_consts
+- Start Date: 2015-03-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow consts declared in functions to depend on function type parameters. Since
+this raises some issues in match and type checks, use an conservative approach
+that allows only minimal use of such consts when a constant expression is
+required.
+
+# Motivation
+
+Consider the following line of code in a non-generic context:
+
+```rust
+let a: [u8; T::N + T::N] = [0u8; 2*T::N];
+```
+
+In the current associated consts implementation (modulo some temporary wrinkles
+involving constants associated to traits), this will readily type-check. Since
+the type `T` must be known in non-generic code, `T::N` will evaluate to some
+value, say `16`, in which case it is clear that both sides of the assignment
+involve arrays of size `32`.
+
+However, in generic code, if `T` is a type parameter, we are asking the type
```

Comment:
I object to this notion. I don't think we're asking the compiler to prove some generic equation. We're asking it to evaluate a kind of a poor man's static_assert at the point of template instantiation. To wit, given:
```rust
trait Foo {
    const VALUE: usize;
}
```
...then:
```rust
fn bar<T: Foo>() {
    let _: [u8; T::VALUE] = [0; 111];
}
```
...is effectively the same as (using C++ syntax for static_assert):
```rust
fn bar<T: Foo>() {
    static_assert(T::VALUE == 111, "Invalid value");
}
```
Or, in more general:
```rust
let _: [u8; x] = [0; y];
```
...is effectively the same as:
```rust
static_assert(x == y, "");
```
...and if either expression `x` or `y` depends on some template parameter, then such static assertion should be deferred to the point of instantiation of the given template.